File: mobile/lib/features/collection/domain/plant.dart
import 'dart:typed_data';

import '../../care/domain/care_type.dart';

enum PlantVisibility {
  private('private', 'Только я'),
  followers('followers', 'Подписчики'),
  public('public', 'Все');

  const PlantVisibility(this.dbName, this.label);

  final String dbName;
  final String label;

  static PlantVisibility fromDb(String value) => values.firstWhere((v) => v.dbName == value);
}

class Location {
  const Location({required this.id, required this.name, this.lightLevel, this.isOutdoor = false});

  final String id;
  final String name;
  final LightLevel? lightLevel;
  final bool isOutdoor;

  factory Location.fromJson(Map<String, dynamic> json) => Location(
        id: json['id'] as String,
        name: json['name'] as String,
        lightLevel: LightLevel.fromDb(json['light_level'] as String?),
        isOutdoor: json['is_outdoor'] as bool? ?? false,
      );
}

class Plant {
  const Plant({
    required this.id,
    required this.nickname,
    this.speciesId,
    this.speciesName,
    this.locationId,
    this.locationName,
    this.lightLevel,
    this.potMaterial,
    this.visibility = PlantVisibility.followers,
    this.notes,
    this.nextWaterAt,
    this.photoUrl,
    this.photoBytes,
  });

  final String id;
  final String nickname;
  final String? speciesId;
  final String? speciesName;
  final String? locationId;
  final String? locationName;
  final LightLevel? lightLevel;
  final PotMaterial? potMaterial;
  final PlantVisibility visibility;
  final String? notes;

  /// Денормализовано из графика полива — для статуса в списке коллекции.
  final DateTime? nextWaterAt;

  /// Обложка: подписанная ссылка из Storage или байты (демо-режим).
  final String? photoUrl;
  final Uint8List? photoBytes;

  /// Путь обложки в бакете plant-photos — чтобы получить подписанную ссылку.
  static String? coverPathOf(Map<String, dynamic> json) => (json['cover'] as Map?)?['storage_path'] as String?;

  /// Ожидает выборку вида
  /// `*, species(latin_name, common_names), locations(name, light_level), care_schedules(type, next_due_at),
  /// cover:plant_photos!plants_cover_photo_fk(storage_path)`.
  factory Plant.fromJson(Map<String, dynamic> json, {String? photoUrl}) {
    final species = json['species'] as Map<String, dynamic>?;
    final location = json['locations'] as Map<String, dynamic>?;
    final schedules = (json['care_schedules'] as List?)?.cast<Map<String, dynamic>>() ?? const [];
    final water = schedules.where((s) => s['type'] == 'water' && s['next_due_at'] != null);
    return Plant(
      id: json['id'] as String,
      nickname: json['nickname'] as String,
      speciesId: json['species_id'] as String?,
      speciesName: species == null ? null : speciesDisplayName(species),
      locationId: json['location_id'] as String?,
      locationName: location?['name'] as String?,
      lightLevel: LightLevel.fromDb(location?['light_level'] as String?),
      potMaterial: PotMaterial.fromDb(json['pot_material'] as String?),
      visibility: PlantVisibility.fromDb(json['visibility'] as String? ?? 'followers'),
      notes: json['notes'] as String?,
      nextWaterAt: water.isEmpty ? null : DateTime.parse(water.first['next_due_at'] as String).toLocal(),
      photoUrl: photoUrl,
    );
  }
}

/// Русское народное название, если есть, иначе латинское.
String speciesDisplayName(Map<String, dynamic> species) {
  final ru = (species['common_names'] as Map?)?['ru'] as List?;
  return ru != null && ru.isNotEmpty ? ru.first as String : species['latin_name'] as String;
}

/// Черновик нового растения с формы добавления.
class NewPlant {
  const NewPlant({
    required this.nickname,
    this.speciesId,
    this.locationId,
    this.potMaterial,
    this.visibility = PlantVisibility.followers,
    this.waterIntervalDays,
    this.lastWateredAt,
    this.notes,
  });

  final String nickname;
  final String? speciesId;
  final String? locationId;
  final PotMaterial? potMaterial;
  final PlantVisibility visibility;

  /// Если null — берётся из базы знаний по виду, иначе 7 дней.
  final double? waterIntervalDays;
  final DateTime? lastWateredAt;
  final String? notes;
}

File: web/src/lib/domain/plant.ts
import type { CareEvent, CareSchedule, LightLevel, PotMaterial } from "./care";

export const VISIBILITIES = {
  private: "Только я",
  followers: "Подписчики",
  public: "Все",
} as const;
export type Visibility = keyof typeof VISIBILITIES;

export interface Location {
  id: string;
  name: string;
  lightLevel: LightLevel | null;
}

export interface Plant {
  id: string;
  nickname: string;
  speciesId: string | null;
  speciesName: string | null;
  speciesSlug: string | null;
  locationId: string | null;
  locationName: string | null;
  lightLevel: LightLevel | null;
  potMaterial: PotMaterial | null;
  visibility: Visibility;
  notes: string | null;
  /** Денормализовано из графика полива — для статуса в списке коллекции. */
  nextWaterAt: Date | null;
  /** Обложка: подписанная ссылка из Storage или data URL (демо-режим). */
  photoUrl: string | null;
  createdAt: Date;
}

export interface PlantDetails {
  plant: Plant;
  schedules: CareSchedule[];
  events: CareEvent[];
}

/** Черновик нового растения с формы добавления. */
export interface NewPlant {
  nickname: string;
  speciesId?: string | null;
  locationId?: string | null;
  potMaterial?: PotMaterial | null;
  visibility?: Visibility;
  /** Если не задан — из базы знаний по виду, иначе 7 дней. */
  waterIntervalDays?: number | null;
  lastWateredAt?: Date | null;
  notes?: string | null;
}

type Row = Record<string, unknown>;

/**
 * Ожидает выборку `*, species(slug, latin_name, common_names), locations(name, light_level),
 * care_schedules(type, next_due_at), cover:plant_photos!plants_cover_photo_fk(storage_path)`.
 */
export function plantFromRow(r: Row, photoUrl: string | null = null): Plant {
  const species = r.species as { slug?: string; latin_name: string; common_names?: { ru?: string[] } } | null;
  const location = r.locations as { name: string; light_level: LightLevel | null } | null;
  const schedules = (r.care_schedules as { type: string; next_due_at: string | null }[] | null) ?? [];
  const water = schedules.find((s) => s.type === "water" && s.next_due_at);
  return {
    id: r.id as string,
    nickname: r.nickname as string,
    speciesId: (r.species_id as string | null) ?? null,
    speciesName: species ? (species.common_names?.ru?.[0] ?? species.latin_name) : null,
    speciesSlug: species?.slug ?? null,
    locationId: (r.location_id as string | null) ?? null,
    locationName: location?.name ?? null,
    lightLevel: location?.light_level ?? null,
    potMaterial: (r.pot_material as PotMaterial | null) ?? null,
    visibility: ((r.visibility as Visibility | null) ?? "followers"),
    notes: (r.notes as string | null) ?? null,
    nextWaterAt: water?.next_due_at ? new Date(water.next_due_at) : null,
    photoUrl,
    createdAt: new Date((r.created_at as string | undefined) ?? Date.now()),
  };
}

export const coverPathOf = (r: Row) => ((r.cover as { storage_path?: string } | null)?.storage_path ?? null);

export type PlantStatus = "ok" | "soon" | "overdue";

/** Статус для точки в углу карточки: просрочено / полив сегодня-завтра / всё хорошо. */
export function plantStatus(p: Pick<Plant, "nextWaterAt">, now: Date): PlantStatus {
  if (!p.nextWaterAt) return "ok";
  const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
  if (p.nextWaterAt < today) return "overdue";
  if (p.nextWaterAt.getTime() < today.getTime() + 2 * 86_400_000) return "soon";
  return "ok";
}
